suite/tests/client-interface/drx-scattergather-instrumentation-fault.dll.c

```diff
@@ -0,0 +1,279 @@
+typedef struct {
+    app_pc app_addr;
+    bool is_scatter_gather;
+    bool do_post_instruction_instrumentation;
+} instru_data_t;
+
+static client_id_t client_id;
+
+static void *faulting_memory;
+
+static dr_signal_action_t
+signal_event(void *drcontext, dr_siginfo_t *info)
+{
+    if (info->sig == SIGSEGV && info->access_address == faulting_memory) {
+        instr_t faulting_instr;
+        instr_init(drcontext, &faulting_instr);
+
+        /* Skip over the faulting instruction. */
+        info->raw_mcontext->pc =
+            decode(drcontext, info->raw_mcontext->pc, &faulting_instr);
+
+        instr_free(drcontext, &faulting_instr);
+        return DR_SIGNAL_SUPPRESS;
+    }
+
+    return DR_SIGNAL_DELIVER;
+}
+
+static void
+insert_faulting_store(void *drcontext, instrlist_t *ilist, instr_t *where,
+                      app_pc app_addr)
+{
+    reg_id_t reg_ptr;
+
+    if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_ptr) !=
+        DRREG_SUCCESS) {
+        DR_ASSERT(false);
+    }
+
+    instrlist_insert_mov_immed_ptrsz(drcontext, (ptr_int_t)faulting_memory,
+                                     opnd_create_reg(reg_ptr), ilist, where, NULL, NULL);
+
+    instrlist_meta_preinsert(
+        ilist, where,
+        INSTR_XL8(XINST_CREATE_store(drcontext, OPND_CREATE_MEMPTR(reg_ptr, 0),
+                                     opnd_create_reg(reg_ptr)),
+                  app_addr));
+
+    if (drreg_unreserve_register(drcontext, ilist, where, reg_ptr) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+}
+
+static void
+handle_post_instruction(void *drcontext, instrlist_t *ilist, instr_t *where,
+                        app_pc app_addr)
+{
+    int i;
+    instr_t *prev_instr = instr_get_prev_app(where);
+    bool seen_memref = false;
+
+    for (i = 0; i < instr_num_dsts(prev_instr); ++i) {
+        const opnd_t dst = instr_get_dst(prev_instr, i);
+        if (opnd_is_memory_reference(dst)) {
+            if (seen_memref) {
+                DR_ASSERT_MSG(false, "Found inst with multiple memory destinations");
+                break;
+            }
+
+            seen_memref = true;
+            insert_faulting_store(drcontext, ilist, where, app_addr);
+        }
+    }
+    for (i = 0; i < instr_num_srcs(prev_instr); ++i) {
+        const opnd_t src = instr_get_src(prev_instr, i);
+        if (opnd_is_memory_reference(src)) {
+            if (seen_memref) {
+                DR_ASSERT_MSG(false, "Found inst with multiple memory destinations");
+                break;
+            }
+
+            seen_memref = true;
+            insert_faulting_store(drcontext, ilist, where, app_addr);
+        }
+    }
+}
+
+static dr_emit_flags_t
+event_app_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                   bool translating, void **user_data)
+{
+    instru_data_t *data = (instru_data_t *)dr_thread_alloc(drcontext, sizeof(*data));
+    data->app_addr = 0;
+    data->is_scatter_gather = false;
```

Comment:
Could we figure this out in instrumentation phase itself using drmg_orig_app_instr_for_fetch?

Reply:
Thanks.